README.md
Michigan EcoData: Don Edwards San Fransisco Bay Wildlife Refuge Climate Modeler
===============================================================================
By Akshey Deokule <aksheyd@umich.edu>

# Introduction
This project is a Climate Modeler for the DESFB, a wildlife refuge 5 minutes away from my home in Fremont, CA which I frequent, using US Wildlife & Refuge data on the park's numerous species. I used C++ for the user environment and Python (Pandas) for the data analysis, sorting, and calculations of the danger level of species depending on the climate variables given.

The calculations are (for now) based on a single variable called DangerValue which is changed depending on classifications of the species given and when it hits a 100 depending on what is changed in the climate modeler, that species is considered extinct in the model. There are many variables which truly affect a ecosystem, and so this is a very simple attempt to try and replicate that. I intend to add more variables in the future.

# Get Started
###### Disclaimer: This project is *NOT* intended to be heavily scientific. The calculations done on the ecosystem's survivability are loosely based on scientific articles and journals (even some Wikipedia). The source for the dataset of the flora and fauna can be found at https://fws.gov/refuge/Don_Edwards_San_Francisco_Bay/ 
###### *Requires Python 3 and C++*

```console 
$ make main.exe
$ ./main.exe
```




CHANGELOG.md
- one more time
- updated readme again lmao
- updated readme
- added changelog
- c++ to python calling works, python script now able to iterate through dataset to calculate danger level of birds
- changed readme
- did so much stuff just to realize that I spelt the filenames wrong :(
- updated makefile, still cant get boost to run or python to run in script or c++ system()
- added a ton of things
- started off the file p much
- read me added
- Initial commit.


main.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <cstdlib>
#include <chrono>
#include <thread>

using namespace std;

void command_screen();
void DESFB_overview();

int main() {
    int input;

    system("clear");
    cout << "Welcome to the Don Edwards San Fransisco Bay Wildlife Refuge Climate Modeler" << endl;
    cout << "Please choose an option from below:" << endl;
    cout << "1: Start" << endl;
    cout << "2: Quit" << endl;

    while (cin >> input) {
        if (input == 1) {
            DESFB_overview();
            break;
        } else if (input == 2) {
            break;
        } else {
            command_screen();
        }
    }
    return 0;
}

void command_screen() {
    cout << "Incorrect input selected, please try again:" << endl;
    cout << "1: Start" << endl;
    cout << "2: Quit" << endl;
}

void DESFB_overview() {
    system("clear");
    int input;

    cout << "Don Edwards San Francisco Bay National Wildlife Refuge, CA, USA" << endl;
    cout << "===============================================================" << endl << endl;

    cout << "Habitats: Marsh, Ponds, Mudflat, Vernal Pools, Uplands" << endl;
    cout << "Species: 269 Birds, 28 Mammals, 12 Amphibian/Reptiles, 62 Fish, 335 Fauna" << endl;
    cout << endl;

    /*
    std::this_thread::sleep_for(chrono::nanoseconds(10));
    std::this_thread::sleep_until(std::chrono::system_clock::now() + std::chrono::seconds(15));
    system("clear");
    */

    cout << "Here's a sneakpeek at some of the birds, mammals, etc" << endl;
    cout << "For the Danger Level coloumn, this number is calculated based on the inital data given" << endl <<
            "from the US Fish & Wildlife Service. It is based on many factors including the species" << endl <<
            "origin, endangered-level, etc." << endl;
    /*
    std::this_thread::sleep_for(chrono::nanoseconds(10));
    std::this_thread::sleep_until(std::chrono::system_clock::now() + std::chrono::seconds(4));
    */

    string filename = "\"/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/sheet_analyzer.py\"";
    string command = "python3 ";
    command += filename;
    system("clear");
    system(command.c_str());

    filename = "\"/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/print_data.py\"";
    command = "python3 ";
    command += filename;
    system("clear");
    system(command.c_str());

    cin >> input;
}

print_data.py
#!/usr/bin/python3

import pandas as pd
import io
import json
import os

bs_df = pd.read_csv (r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/output_birds.csv')
bs_df = bs_df[:270] 
print("Birds: ")
print("======")
print(bs_df[["Common Name", "Scientific Name", "Danger Level"]].head())
print("Max Danger Level for Birds = ", bs_df["Danger Level"].max())
print("Min Danger Level for Birds = ", bs_df["Danger Level"].min())
print("Average Danger Level for Birds = ", "{:.2f}".format(bs_df["Danger Level"].mean()))
print("Amount of Birds = ", len(bs_df))
os.remove("output_birds.csv")
print()

m_df = pd.read_csv(r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/output_mammals.csv')
m_df = m_df[:29] 
print("Mammals: ")
print("======== ")
print(m_df[["Common Name", "Scientific Name", "Danger Level"]].head())
print("Max Danger Level for Mammals = ", m_df["Danger Level"].max())
print("Min Danger Level for Mammals = ", m_df["Danger Level"].min())
print("Average Danger Level for Mammals = ", "{:.2f}".format(m_df["Danger Level"].mean()))
print("Amount of Mammals = ", len(m_df))
os.remove("output_mammals.csv")
print()

ar_df = pd.read_csv(r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/output_amphibianreptiles.csv')
ar_df = ar_df[:14] 
print("Amphibians/Reptiles: ")
print("==================== ")
print(ar_df[["Common Name", "Scientific Name", "Danger Level"]].head())
print("Max Danger Level for A/R = ", m_df["Danger Level"].max())
print("Min Danger Level for A/R = ", m_df["Danger Level"].min())
print("Average Danger Level for A/R = ", "{:.2f}".format(ar_df["Danger Level"].mean()))
print("Amount of A/R = ", len(ar_df))
os.remove("output_amphibianreptiles.csv")
print()

f_df = pd.read_csv(r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/output_fishs.csv')
f_df = f_df[:58] 
print("Fish: ")
print("===== ")
print(f_df[["Common Name", "Scientific Name", "Danger Level"]].head())
print("Max Danger Level for Fish = ", f_df["Danger Level"].max())
print("Min Danger Level for Fish = ", f_df["Danger Level"].min())
print("Average Danger Level for Fish = ", "{:.2f}".format(f_df["Danger Level"].mean()))
print("Amount of Fish = ", len(f_df))
os.remove("output_fishs.csv")
print() 

sheet_analyzer.py
#!/usr/bin/python3

import pandas as pd
import io
import json

"""
# Adds command line arguements to take in input for climate
import argparse
parser = argparse.ArgumentParser(description='Analyzer for data')
parser.add_argument("-p", "--print_string", help="Takes in the supplied argument for calculation.", nargs='*')
args = parser.parse_args()
print(args.print_string)
"""

# Birds CSV
bs_df = pd.read_csv (r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/BirdSheet.csv')
bs_df = bs_df[:270] 

# Mammals CSV
m_df = pd.read_csv (r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/MammalsSheet.csv')
m_df = m_df[:30] 

# Amphibians/Reptiles CSV
ar_df = pd.read_csv (r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/AmphibianReptilesSheet.csv')
ar_df = ar_df[:14] 

# Reptiles CSV
f_df = pd.read_csv (r'/Users/aksheydeokule/Documents/EcoData S2/Climate Modeler/FishsSheet.csv')
f_df = f_df[:58] 

#converts Occurrence column into string for if statements in DangerLevel func
bs_df["Occurrence"] = bs_df["Occurrence"].astype(str)
bs_df["Classification"] = bs_df["Classification"].astype(str)
bs_df["Federal"] = bs_df["Federal"].astype(str)
bs_df["State"] = bs_df["State"].astype(str)

m_df["Occurrence"] = m_df["Occurrence"].astype(str)
m_df["Classification"] = m_df["Classification"].astype(str)
m_df["Federal"] = m_df["Federal"].astype(str)
m_df["State"] = m_df["State"].astype(str)

ar_df["Occurrence"] = ar_df["Occurrence"].astype(str)
ar_df["Classification"] = ar_df["Classification"].astype(str)
ar_df["Federal"] = ar_df["Federal"].astype(str)
ar_df["State"] = ar_df["State"].astype(str)

f_df["Occurrence"] = f_df["Occurrence"].astype(str)
f_df["Classification"] = f_df["Classification"].astype(str)
f_df["Federal"] = f_df["Federal"].astype(str)
f_df["State"] = f_df["State"].astype(str)

def DangerLevel (row1):
    returnVal = 0

    # Locally nested birds have acclimated to the climate better
    if (row1["Occurrence"] == "nests locally"):
        returnVal += 1
    # This is just an estimate between accidental occurring birds and locally nested
    elif (row1["Occurrence"] == "nan"):
        returnVal += 5
    # Accidental birds have no evolutionary acclimation to new climate and are more in danger
    elif (row1["Occurrence"] == "accidental"):
        returnVal += 10


    # Solo habitat lowers survival chance (LOOK INTO CHANGING CLIMATE BASED ON HABTIAT)
    if ((row1["Occurrence"] == "upland") or (row1["Occurrence"] == "marsh") or (row1["Occurrence"] == "tidal sloughs")):
        returnVal += 10
    # Multiple habitats means greater survival in more areas
    elif ((row1["Occurrence"] == "all habitats")):
        returnVal += 5


    # Solo habitat lower survival chance (LOOK INTO CHANGING CLIMATE BASED ON HABTIAT))
    if (row1["Occurrence"] == "present in tidal waters"):
        returnVal += 10
    # More habitats -> greater survival
    elif (row1["Occurrence"] == "present in tidal waters and ponds"):
        returnVal += 5
    # Rare occurence means lower population and lower survival
    elif (row1["Occurrence"] == "rare in tidal waters"):
        returnVal += 20


    # Native species, like locally nested, have better climate acclimation
    if (row1["Classification"] == "native"):
        returnVal += 1
    # Non Native species are given higher danger because of above 
    elif (row1["Classification"] == "non-native"):
        returnVal += 5

    # Endangered species are close to extinction on a Federal level
    if (row1["Federal"] == "E"):
        returnVal += 50
    # Threatened species are close to becoming endangered
    elif (row1["Federal"] == "T"):
        returnVal += 25
    
    # Every state endangered species is also federally endangered
    if (row1["State"] == "E"):
        returnVal += 0
    # Threatened species are close to becoming endangered
    elif (row1["State"] == "T"):
        returnVal += 25
    # SSC/SC (Species of Special Concern) is a CA specific classification for threatened and declining population species
    elif (row1["State"] == "SSC") or (row1["State"] == "SC"):
        returnVal += 50
    # SP means State Protected 
    elif ((row1["State"] == "SP") or (row1["State"] == "FP")):
        return 0

    return returnVal

# Creates Danger Level column
bs_df.insert(len(bs_df.columns), 'Danger Level', 0)
m_df.insert(len(m_df.columns), 'Danger Level', 0)
ar_df.insert(len(ar_df.columns), 'Danger Level', 0)
f_df.insert(len(f_df.columns), 'Danger Level', 0)

# Iterates through rows to calculate Danger Level
bs_df['Danger Level'] = bs_df.apply(lambda row1: DangerLevel(row1), axis = 1)
m_df['Danger Level'] = m_df.apply(lambda row1: DangerLevel(row1), axis = 1)
ar_df['Danger Level'] = ar_df.apply(lambda row1: DangerLevel(row1), axis = 1)
f_df['Danger Level'] = f_df.apply(lambda row1: DangerLevel(row1), axis = 1)

# Turns all datasets into csv files
bs_df.to_csv("output_birds.csv", index = False)
m_df.to_csv("output_mammals.csv", index = False)
ar_df.to_csv("output_amphibianreptiles.csv", index = False)
f_df.to_csv("output_fishs.csv", index = False)